### ISSUE TITLE
Flaky LSP HoverRequest response

### ISSUE THREAD

**Issue by marvinborner** [effekt-lang/effekt#366]
While working on the LSP testing suite (#351), I had several problems with the [textDocument/hover](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_hover) request. I initially believed the problem to be in my LSP client implementation (and I was correct, to some extent) but now I think the problem also lies in Effekt/Kiama.

A reproducible communication flow that sometimes returns the wrong result is as follows:

1. `InitializeRequest` -> ...
2. `InitializedNotification`
3. `DidOpenTextDocumentNotification` `effect Exc(msg: String): Unit`
4. `HoverRequest` (e.g. line 0, char 8) -> `null` / { ... }

Where the last `HoverRequest` returns `null` instead of the correct result around 1/2 of the times.

You can try it yourself using [this JS script](https://gist.github.com/marvinborner/ddc72b214352a485d308a7b1c157230b).

I spent a lot of time trying to debug this but ultimately did not get much further than that this function call sometimes returns `None` instead of `Some(...)`:

https://github.com/effekt-lang/effekt/blob/e0d28ef099d3004ddfdde63d1e07826c7983b265/effekt/jvm/src/main/scala/effekt/Server.scala#L91

I was not able to reproduce this issue with other similar request types.

**Comment by b-studios** [comment]
Maybe it is some kind of timing issue? It looks like `process` (which in turn will compile the file) will only be called `onsave`

https://github.com/effekt-lang/kiama/blob/3401470a4cbbe294d2df0c6923e919b8a45ca1f5/jvm/src/main/scala/kiama/util/Server.scala#L498

so maybe this call is not fully processed, when the onhover is triggered?

**Comment by b-studios** [comment]
I don't know what's the LSP best practices here. Should there be _some_ form of synchronization? Should we block until the first `process` is completed? Or is this completely against how LSP usually works?

**Comment by marvinborner** [comment]
Don't we already process on `textDocument/didOpen` [here](https://github.com/effekt-lang/kiama/blob/3401470a4cbbe294d2df0c6923e919b8a45ca1f5/jvm/src/main/scala/kiama/util/Server.scala#L486) and, in the reproduction, [here](https://gist.github.com/marvinborner/ddc72b214352a485d308a7b1c157230b#file-repro-js-L41)?

I assumed `process` to be blocking already, otherwise it should probably return a future or something, right? I think this is probably the best practice, I have not come across synchronization yet.

It's also interesting that if you comment out [this line](https://gist.github.com/marvinborner/ddc72b214352a485d308a7b1c157230b#file-repro-js-L66), i.e. always open and close the same file, the following hover results will always stay on the first result (`null`/`{...}`).

**Comment by b-studios** [comment]
I just thought there might be a race condition with save, open, or another action that will indeed perform the processing but is not finished yet.

**Comment by marvinborner** [comment]
I tried the fix we talked about again; doesn't change anything. I believe that was the idea, right:

``` scala
def getSymbolAt(position: Position)(implicit C: Context): Option[(Tree, Symbol)] =
  for {
    id <- getIdTreeAt(position)
    _ = C.compiler.Frontend.run(position.source) // hotfix
    sym <- C.symbolOption(id)
  } yield (id, resolveCallTarget(sym))
```

**Comment by b-studios** [comment]
Yes, exactly. This is really strange. Does adding `sleep` in some places help to diagnose?

**Comment by marvinborner** [comment]
Added many sleeps, doesn't change anything (except execution time)... So weird

**Comment by timsueberkrueb** [comment]
@marvinborner could you maybe come up with a [unit test](https://github.com/effekt-lang/effekt/pull/885/files#diff-82dd325639594112ab3c399488471125d97d71d9fb70bf92d9f368099ce49fb7) that checks if this is fixed in the [new language server implementation](https://github.com/effekt-lang/effekt/pull/885)?
I didn't see this problem so far, neither in the old nor the new implementation.

**Comment by marvinborner** [comment]
The issue seems to be fixed! I can't reproduce the flakiness using my script (linked above) anymore. Interestingly, as you mentioned, even the old implementation doesn't seem to have this problem anymore. We could try bisecting to find a specific flakiness source, but aside from running the test a bunch of times and comparing the results, I don't see any way of preventing regressions in the future using a unit test.

**Comment by timsueberkrueb** [comment]
Awesome, I guess this can be closed then.
